### PR TITLE
bug fix for polynomial domain serialization

### PR DESCRIPTION
**Pull request by nden**
This is the part of #9941 that can go in 4.0.1. It fixes the serialization of polynomials and updates the documentation. Once this is merged #9941 will be updated and merged into master with a milestone of 4.1.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -366,6 +369,9 @@ astropy.io.votable
 astropy.modeling
 ^^^^^^^^^^^^^^^^
 
+- Improvements to documentation clearing up how domain and window attributes are
```
Comment (bsipocz) [inline CHANGES.rst]: no need to have an entry for the documentation

File: astropy/modeling/polynomial.py [hunk 2]
```diff
@@ -141,6 +141,10 @@ class OrthoPolynomialBase(PolynomialBase):
 
     The polynomials implemented here require a maximum degree in x and y.
 
+    For explanation of `x_domain`, `y_domain`, `x_window` and `y_window`
```
Comment (bsipocz) [inline astropy/modeling/polynomial.py]: parameter names, here and below, need to be double backticked, otherwise it'll cause the sphinx build fails.

Comment (nden) [reply astropy/modeling/polynomial.py]: Thanks @bsipocz, all fixed now.

**Comment by jehturner** [comment]
For what it's worth, my `gwcs` serialization test passes with this branch and with `master` & `4.0.x` since it has been merged. I had trouble testing against an older version of `asdf`, but I understand that you have checked that. Thanks!

**Comment by bsipocz** [comment]
@jehturner - just to confirm, you had troubles with older asdf version on the 4.0.x branch, too?

**Comment by bsipocz** [comment]
(CI seems to be happy there, but if I can image it's missing something that you run into locally). So if that's the case, please open an issue as we shouldn't break stuff that used to work on the bugfix branch (if it hasn't worked before either, then it's fine I suppose).

**Comment by jehturner** [comment]
No, I had a bit of trouble getting a sensible combination working together to get a useful result. Now that this is merged into 4.0.x it should be much easier, but simply installing the conda package was going to muck up my env, but I can try a bit harder if you want :slightly_smiling_face:.

**Comment by bsipocz** [comment]
I suppose it's fine, but next time if you run into it, using release asdf, and 4.0.x, then please do open the issue.

**Comment by jehturner** [comment]
OK, I'm a bit confused. The minimum asdf version was already changed in the `4.0.x` branch on 2 January, from `2.3.0` to `2.5.0` and -- I think as a result -- all the transform tests simply get skipped when `4.0.x` is installed with the latest `asdf 2.4.2` from astroconda. And if I revert the minimum asdf version in `__init__.py` to get them to run then I get another schema file warning about `ortho_polynomial-1.0.0` from the two new tests (causing them to fail). But my `gwcs` serialization test does pass with this combination (also with a warning about `ortho_polynomial-1.0.0`), so apparently it does work...

**Comment by bsipocz** [comment]
I can't comment on astroconda. My understanding behind the previous version bump is that the previous minimum was missing the schemas and this was totally incompatible with 4.0.

**Comment by pllim** [comment]
Re: astroconda -- Whoever made the `asdf` releases has not been updating its version in `astroconda`. Something like astroconda/astroconda-contrib#610 needs to happen. cc @jdavies-st and @eslavich

**Comment by jehturner** [comment]
The thing is, @bsipocz, `astropy 4.0` went out with the requirement `asdf>=2.3` and now the unpublished `4.0.x` branch already has `asdf>=2.5` due to whatever the missing schemas problem was, so why not just make it `2.5.2` and fix this problem definitively? Is there really an advantage to requiring `2.5.0` instead of `2.5.2` when it's not even the existing constraint and the latter is a minor update? Just saying (sorry @nden for dragging this out even more, but it seems worth asking) :wink:.

**Comment by jehturner** [comment]
This assumes that `asdf 2.5.2` could be put out in time for `astropy 4.0.1`, of course.

**Comment by bsipocz** [comment]
Ok, that's a fair point that the 2.5 requirement hasn't been released yet, so yes I don't see a blocking there to bump it up more. I'm still uncomfortable with the first bump, but understand it's necessary.

**Comment by jehturner** [comment]
Good, so I suppose @nden & @perrygreenfield can comment on whether they plan to put out `2.5.2` soon enough and whether I can help in some way. I think updating the `asdf` constraints in `__init__.py` and `setup.cfg` would be sufficient in combination with the existing changes from this PR to eliminate the fallbacks & associated warnings?

**Comment by jdavies-st** [comment]
The dependency of `astropy 4.0.x` should be `asdf 2.5.0`.  The latest patch `2.5.1` was a minor packaging fix that only affected testing on installed versions of the package and was needed for linux distributions of the package.  And I don't think anything currently in `dev` is needed for `astropy` right now.

I'll add that the only reason `astropy 4.0.0` needs `asdf 2.5.0` is because the schemas for seralizing Astropy classes are in the `asdf-standard` git submodule of `asdf`.  Because of this, one gets a chicken/egg problem when doing releases that one package will have the incorrect dependencies (in this case `astropy`) when released.

Alternatively, one could package the schemas that `astropy` needs that are currently in `asdf-standard` in a different package, say `asdf-astronomy`, so that it is separate and doesn't depend on releasing `asdf` to get updated functionality.

**Comment by jdavies-st** [comment]
Also, `asdf 2.5.1` is now available via `astroconda` channel.  Thanks @rendinam.

Apparently it is also available from `conda-forge`, as they have a bot that sort-of automates updating the recipe (though doesn't automate the build dependencies).

**Comment by jehturner** [comment]
> And I don't think anything currently in dev is needed for astropy right now.

But my understanding is that there are newer schema files in `dev` that are needed for the changes in this PR to function as intended without having to fall back to older versions and suppress the resulting warnings.

**Comment by jehturner** [comment]
Eg. See https://github.com/astropy/astropy/pull/9941#discussion_r379633960.

**Comment by nden** [comment]
@bsipocz @jehturner To close the loop on this - we are going to have a 2.5.2 release of asdf in the next few days. Given that the requirement is `>=2.5` and this will pick up `2.5.2` do we need to bump the version in astropy?

**Comment by bsipocz** [comment]
Given we haven't yet released the `>=2.5` requirement here, I would say bumping it to `>=2.5.2` would be the cleaner approach. You're absolutely right that 2.5.2 will be picked up for new installs, but we can still end up with corner cases when someone has 2.5 and runs into errors.

**Comment by jehturner** [comment]
Either way seems reasonable to me -- this PR **should** turn the corner cases @bsipocz mentions into warnings rather than errors, but requiring `>=2.5.2` could be even safer.

**Comment by nden** [comment]
I'll update the minimum version to 2.5.2 once the asdf release is out.

**Pull request merged**